### PR TITLE
refactor: Change Deref target of KeyPrefix and KeyName to str

### PR DESCRIPTION
**Pull request by Techassi**
Came up in https://github.com/stackabletech/hdfs-operator/pull/454.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 
 ## [Unreleased]
 
+### Changed
+
+- Change Deref target of `KeyPrefix` and `KeyName` to `str` ([#725]).
```
Comment (sbernauer) [inline CHANGELOG.md:9]:
```suggestion
- Change Deref target of `KeyPrefix` and `KeyName` from `String` to `str` ([#725]).
```

**Review by NickLarsenNZ — DISMISSED** [review]
LGTM - only thing that is arguably missing would be a test to show how it gets used (eg: https://github.com/stackabletech/hdfs-operator/pull/454#discussion_r1462961182)

**Comment by Techassi** [comment]
Sure, I can add a few tests which should cover the use-case you mentioned in `hdfs-operator`.

**Review by NickLarsenNZ — APPROVED** [review]

**Review by sbernauer — DISMISSED** [review]

File: src/kvp/key.rs [hunk 2]
```diff
@@ -399,4 +402,30 @@ mod test {
...
+    #[rstest]
+    #[case("app.kubernetes.io/name", true)]
+    #[case("name", false)]
+    fn key_prefix_deref(#[case] key: &str, #[case] expected: bool) {
+        let label = Label::try_from((key, "zookeeper")).unwrap();
+
+        let is_valid = label
+            .key()
+            .prefix()
+            .is_some_and(|prefix| *prefix == "app.kubernetes.io");
```
Comment (NickLarsenNZ) [inline src/kvp/key.rs:415]: Shouldn't this be:

```rust
.prefix()
.as_deref() == "app.kubernetes.io";
```

As Nat was suggesting in the coomment linked in the PR review?

Or maybe I'm misunderstanding what she meant.

**Review by NickLarsenNZ — APPROVED** [review]
LGTM

File: src/kvp/key.rs (hunk 2 above)
Comment (NickLarsenNZ) [reply src/kvp/key.rs:415]: Yeah, so the `as_deref()` comparison to a `&str` doesn't work (as per your linked comment), but I think this is still a net-positive improvement.

Comment (Techassi) [reply src/kvp/key.rs:415]: Mark as resolved for now, tracking done in https://github.com/stackabletech/operator-rs/issues/727

**Pull request merged**